lib/functional/iter_extras.nit

```diff
@@ -0,0 +1,370 @@
+
+        # Generates an `Iterator` whose elements are sorted by the function
+        # passed in argument.
+        #
+        # ~~~~nitish
+        # class Person
+        #       var name: String
+        # end
+        #
+        # var ps = [new Person("Turing"), new Person("Curry"), new Person("Alfredo")]
+        # var ordered_names = ps.iterator.order_by(&Person::name).map(&Person::name).to_a
+        # assert ordered_names == ["Alfredo", "Curry", "Turing"]
+        # ~~~~
+        fun order_by(f: Fun1[E, Comparable]): OrderedIter[E]
+        do
+                return new OrderedIter[E](self, f)
+        end
+
+end
+
+# Base class for all iterators using functional types.
+private abstract class FunIter[OLD,NEW]
+        super Iterator[NEW]
+        var my_iter: Iterator[OLD]
+
+        redef fun next
+        do
+                my_iter.next
+        end
+
+        redef fun start
+        do
+                my_iter.start
+        end
+
+        redef fun finish
+        do
+                my_iter.finish
+        end
+
+        redef fun is_ok
+        do
+                return my_iter.is_ok
+        end
+end
+
+
+class MapIter[A,B]
+        super FunIter[A,B]
+        var f: Fun1[A, B]
+
+        redef fun item
+        do
+                return f.call(my_iter.item)
+        end
+
+end
+
+class EnumerateIter[E]
+        super FunIter[E, Pair[Int,E]]
+
+        redef fun item
+        do
+               return new Pair[Int,E](0, my_iter.item)
+        end
+end
+
+class FilterIter[E]
+        super FunIter[E,nullable E]
+        var pred: Fun1[E, Bool]
+
+        redef init
+        do
+                super
+                if is_ok and not pred.call(my_iter.item) then
+                        next
+                end
+        end
+
+        redef fun item
+        do
+                assert is_ok
+                return my_iter.item
+        end
+
+
+        redef fun next
+        do
+                loop
+                        my_iter.next
+                        if not is_ok then
+                                break
+                        end
+                        var x = my_iter.item
+                        if pred.call(x) then
+                                break
+                        end
+                end
+        end
+end
+
+class FlatMapIter[E]
+        super FunIter[E,Object]
+        var f: Fun1[E, Iterator[Object]]
+        var buffer: Iterator[Object] is noinit
+
+        redef init
+        do
+                super
+                buffer = f.call(my_iter.item)
+        end
+
+        redef fun item
+        do
+                return buffer.item
+        end
+
+        redef fun next
+        do
+                buffer.next
+                if not buffer.is_ok then
+                        super
+                        if is_ok then
+                                buffer = f.call(my_iter.item)
+                        end
+                end
+        end
+end
+
+
+class OrderedIter[E]
```

Comment:
this one is not like the other since the whole initial iterator will be iterated before working.
Maybe add the service on `Array` instead of `Iterator` ?

Reply:
Indeed, this one is consume eargly. However, if you add it only to Array, you restrict yourself to a single type. We couldn't reuse it for other kind of collections. It makes sense too to have an ordered iter after a big chain:

`var xs = ... a random collection ...`
`xs.iterator.map(f1).map(f2).map(f3).order_by(pred)`

Reply:
OK, but could you add `sort_with(f: Fun1[E, Comparable])` to `Array[E]` then implement `Iterator::order_by` using it?